crates/stackable-telemetry/src/instrumentation/axum/mod.rs

```diff
@@ -0,0 +1,468 @@
+            server.address = Empty,
+            server.port = Empty,
+            client.address = Empty,
+            client.port = Empty,
+            http.route = Empty,
+            http.response.status_code = Empty,
+            // TODO (@Techassi): Add network.protocol.version
+        );
+
+        // Set the parent span based on the extracted context
+        debug!("set parent span based on extracted context");
+        let parent = HeaderExtractor::new(req.headers()).extract_context();
+        span.set_parent(parent);
+
+        if let Some(user_agent) = req.user_agent() {
+            span.record("user_agent.original", user_agent);
+        }
+
+        // Setting server.address and server.port
+        // See https://opentelemetry.io/docs/specs/semconv/http/http-spans/#setting-serveraddress-and-serverport-attributes
+
+        if let Some(server_socket_address) = req.server_socket_address() {
+            span.record("server.address", server_socket_address.ip().to_string())
+                .record("server.port", server_socket_address.port());
+        }
+
+        // Setting fields according to the HTTP server semantic conventions
+        // See https://opentelemetry.io/docs/specs/semconv/http/http-spans/#http-server-semantic-conventions
+
+        if let Some(client_socket_address) = req.client_socket_address() {
+            span.record("client.address", client_socket_address.ip().to_string());
+
+            if opt_in {
+                span.record("client.port", client_socket_address.port());
+            }
+        }
+
+        // Only include the headers if the user opted in, because this might
+        // potentially be an expensive operation when many different headers
+        // are present. The OpenTelemetry spec also marks this as opt-in.
+
+        // NOTE (@Techassi): Currently, tracing doesn't support recording
+        // fields which are not registered at span creation which thus makes
+        // it impossible to record request headers at runtime.
+        // See: https://github.com/tokio-rs/tracing/issues/1343
+
+        // FIXME (@Techassi): Add support for this when tracing allows dynamic
+        // fields.
+        // if opt_in {
+        //     span.add_header_fields(req.headers())
+        // }
+
+        if let Some(http_route) = req.matched_path() {
+            span.record("http.route", http_route.as_str());
+        }
+
+        span
+    }
+
+    fn add_header_fields(&self, headers: &HeaderMap) {
+        for (header_name, header_value) in headers {
+            // TODO (@Techassi): Add an allow list for header names
+            // TODO (@Techassi): Handle multiple headers with the same name
+
+            // NOTE (@Techassi): header_name.as_str() always returns lowercase
+            // strings and thus we don't need to call to_lowercase on it.
+            let header_name = header_name.as_str();
+            let field_name = format!("http.request.header.{header_name}");
```

Comment:
The NOTE can be removed? The other TODOs / FIXMEs are for another PR i assume @Techassi ?

Reply:
I can remove the `NOTE` prefix, but I would like to keep the comment in there so that we don't stumble over this in the future.

Reply:
Removed in 5f47aad